File: archive/c/c-sharp/LongestCommonSubsequence.cs
using System.Collections.Generic;

if (
    args is not [var raw1, var raw2]
    || !TryParseList(raw1.AsSpan(), out var a)
    || !TryParseList(raw2.AsSpan(), out var b)
)
{
    Console.WriteLine(
        """
Usage: please provide two lists in the format "1, 2, 3, 4, 5"
"""
    );
    return;
}

Console.WriteLine(string.Join(", ", LCS(a, b)));

static bool TryParseList(ReadOnlySpan<char> span, out List<int> numbers)
{
    numbers = new(span.Count(',') + 1);

    while (!span.IsEmpty)
    {
        int comma = span.IndexOf(',');
        var token = comma >= 0 ? span[..comma] : span;

        span = comma >= 0 ? span[(comma + 1)..] : [];

        if (!int.TryParse(token, out int n))
            return false;

        numbers.Add(n);
    }

    return true;
}

static List<int> LCS(List<int> a, List<int> b)
{
    int n = a.Count, m = b.Count;
    if (n == 0 || m == 0) return [];

    int[,] dp = new int[n + 1, m + 1];

    for (int i = 1; i <= n; i++)
    for (int j = 1; j <= m; j++)
        dp[i, j] = a[i - 1] == b[j - 1]
            ? dp[i - 1, j - 1] + 1
            : Math.Max(dp[i - 1, j], dp[i, j - 1]);

    var result = new List<int>(dp[n, m]);

    int i2 = n, j2 = m;

    while (i2 > 0 && j2 > 0)
    {
        if (a[i2 - 1] == b[j2 - 1])
        {
            result.Add(a[i2 - 1]);
            i2--;
            j2--;
        }
        else if (dp[i2 - 1, j2] >= dp[i2, j2 - 1])
        {
            i2--;
        }
        else
        {
            j2--;
        }
    }

    result.Reverse();
    return result;
}

File: archive/v/visual-basic/longest-common-subsequence.vb
Imports System
Imports System.Collections.Generic

Public Module LongestCommonSubsequence

    Public Sub Main(args As String())

        If args.Length < 2 Then
            ShowUsage()
            Return
        End If

        Dim list1 = Parse(args(0))
        Dim list2 = Parse(args(1))

        If list1 Is Nothing OrElse list2 Is Nothing Then
            ShowUsage()
            Return
        End If

        Dim result = LCS(list1, list2)

        Console.WriteLine(String.Join(", ", result))

    End Sub


    Private Function LCS(list1 As List(Of String),
                         list2 As List(Of String)) As List(Of String)

        If list1.Count = 0 OrElse list2.Count = 0 Then
            Return New List(Of String)
        End If

        If list1(0) = list2(0) Then

            Dim rest = LCS(list1.GetRange(1, list1.Count - 1),
                           list2.GetRange(1, list2.Count - 1))

            rest.Insert(0, list1(0))
            Return rest

        End If

        Dim option1 = LCS(list1, list2.GetRange(1, list2.Count - 1))
        Dim option2 = LCS(list1.GetRange(1, list1.Count - 1), list2)

        Return If(option1.Count >= option2.Count, option1, option2)

    End Function


    Private Function Parse(input As String) As List(Of String)

        If String.IsNullOrWhiteSpace(input) Then
            Return Nothing
        End If

        Dim result As New List(Of String)

        For Each part In input.Split(","c, StringSplitOptions.RemoveEmptyEntries)
            result.Add(part.Trim())
        Next

        Return result

    End Function


    Private Sub ShowUsage()
        Console.WriteLine(
            "Usage: please provide two lists in the format ""1, 2, 3, 4, 5"""
        )
        Environment.Exit(1)

    End Sub

End Module